src/linear_algebra/pi_tensor_product.lean

```diff
@@ -0,0 +1,427 @@
+    (λ z f i r, by simp [φ.map_smul, smul_smul, mul_comm])
+
+lemma lift_aux_tprod (φ : multilinear_map R s E) (f : Π i, s i) : lift_aux φ (tprod R f) = φ f :=
+by simp only [lift_aux, lift_add_hom, tprod, tprod_coeff, free_add_monoid.lift_eval_of,
+              one_smul, add_con.lift_mk']
+
+lemma lift_aux_tprod_coeff (φ : multilinear_map R s E) (z : R) (f : Π i, s i) :
+  lift_aux φ (tprod_coeff R z f) = z • φ f :=
+by simp [lift_aux, lift_add_hom, tprod_coeff, free_add_monoid.lift_eval_of]
+
+lemma lift_aux.smul {φ : multilinear_map R s E} (r : R) (x : ⨂[R] i, s i) :
+  lift_aux φ (r • x) = r • lift_aux φ x :=
+begin
+  refine pi_tensor_product.induction_on' x _ _,
+  { intros z f,
+    simp [smul_tprod_coeff' r z f, lift_aux_tprod_coeff, lift_aux_tprod_coeff, smul_smul] },
+  { intros z y ihz ihy,
+    rw [smul_add, (lift_aux φ).map_add, ihz, ihy, (lift_aux φ).map_add, smul_add] }
+end
+
+/-- Constructing a linear map `(⨂[R] i, s i) → E` given a `multilinear_map R s E` with the
+property that its composition with the canonical `multilinear_map R s E` is
+the given multilinear map `φ`. -/
+def lift : (multilinear_map R s E) ≃+ ((⨂[R] i, s i) →ₗ[R] E) :=
+{ to_fun := λ φ, { map_smul' := lift_aux.smul, .. lift_aux φ },
+  inv_fun := λ φ', φ'.comp_multilinear_map (mk R s),
+  left_inv := λ φ, by { ext, simp [lift_aux_tprod, linear_map.comp_multilinear_map] },
+  right_inv := λ φ, by { ext, simp [lift_aux_tprod] },
+  map_add' := λ φ₁ φ₂, by { ext, simp [lift_aux_tprod] } }
+
+variables {φ : multilinear_map R s E}
+
+@[simp] lemma lift.tprod (f : Π i, s i) : lift φ (tprod R f) = φ f := lift_aux_tprod φ f
+
+@[simp] lemma lift.tprod' (f : Π i, s i) : (lift φ).1 (tprod R f) = φ f := lift.tprod _
+
+theorem lift.unique {φ' : (⨂[R] i, s i) →ₗ[R] E} (H : ∀ f, φ' (tprod R f) = φ f) :
+  φ' = lift φ :=
+ext $ λ f, by rw [H, lift.tprod]
+
+theorem lift_mk : lift (mk R s) = linear_map.id :=
+eq.symm $ lift.unique $ λ _, rfl
+
+end multilinear
+
+end pi_tensor_product
+
+end semiring
+
+section ring
+namespace pi_tensor_product
+
+open pi_tensor_product
+open_locale tensor_product
+
+variables {ι : Type*} {R : Type*} [comm_ring R]
+variables {s : ι → Type*} [∀ i, add_comm_group (s i)] [∀ i, module R (s i)]
+variables {E : Type*} [add_comm_group E] [semimodule R E]
+
+/- Unlike for the binary tensor product, we require `R` to be a `comm_ring` here, otherwise
+this is false in the case where `ι` is empty. -/
+instance : add_comm_group (⨂[R] i, s i) := semimodule.add_comm_monoid_to_add_comm_group R
+
+lemma neg_tprod_coeff (z : R) (f : Π i, s i) : tprod_coeff R (-z) f = -tprod_coeff R z f :=
+by rw [←neg_one_smul R, ←smul_tprod_coeff', neg_one_smul R]
+
+lemma neg_tprod (f : Π i, s i) (i : ι) : tprod R (update f i (-(f i))) = -tprod R f :=
+by rw [←neg_one_smul R, ←smul_tprod, neg_one_smul R]
+
+lemma neg_tprod' (f : Π i, s i) (i : ι) (m : s i) :
+  tprod R (update f i (-m)) = -tprod R (update f i m) :=
+by rw [←neg_one_smul R, ←smul_tprod', neg_one_smul R]
```

Comment:
I think this is
```suggestion
mk.map_neg f i m
```

Reply:
We need to specify `R` and `s`, but yes.